src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -1135,6 +1093,102 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
+                    cmt = inner_cmt;
+                }
+                if let Categorization::Local(local_id) = cmt.cat {
+                    let parent = self.tcx.map.get_parent_node(local_id);
+                    let may_be_function = match self.tcx.map.get(parent) {
+                        hir::map::NodeItem(item) =>
+                            if let hir::ItemFn(ref function, ..) = item.node {
+                                Some(function)
+                            } else {
+                                None
+                            },
+                        hir::map::NodeImplItem(item) =>
+                            if let hir::ImplItemKind::Method(ref sig, ..) = item.node {
+                                Some(&sig.decl)
+                            } else {
+                                None
+                            },
+                        hir::map::NodeTraitItem(item) =>
+                            if let hir::MethodTraitItem(ref sig, ..) = item.node {
+                                Some(&sig.decl)
+                            } else {
+                                None
+                            },
+                        _ => None
+                    };
+
+                    if let Some(function) = may_be_function {
+                        if let Some(ref arg) = function.inputs.iter()
+                            .find(|ref arg| arg.pat.id == local_id && arg.to_self().is_none()) {
+                            if let hir::TyRptr(Some(lifetime), hir::MutTy { ref ty, ..}) = arg.ty
+                                .node {
+                                if let Ok(snippet) = self.tcx.sess.codemap()
+                                    .span_to_snippet(ty.span) {
+                                    if let Ok(lifetime_snippet) = self.tcx.sess.codemap()
+                                        .span_to_snippet(lifetime.span) {
+                                            db.span_label(arg.ty.span,
+                                                          &format!("use `{} mut {}`",
+                                                                   lifetime_snippet, snippet));
+                                    }
+                                }
+                            }
+                            else if let Ok(snippet) = self.tcx.sess.codemap()
+                                .span_to_snippet(arg.ty.span) {
+                                if snippet.starts_with("&") {
+                                    db.span_label(arg.ty.span,
+                                                  &format!( "use `{}` here to make mutable",
+                                                            snippet.replace("&", "&mut ")));
+                                }
+                            }
+                        }
+                    }
+                }
+                if let Categorization::Local(local_id) = err.cmt.cat {
```

Comment:
As an aside, I think it would be to pull the bodies of these ifs into helper methods with a nice name, like `fn give_suggestion_to_make_referent_mutable()` (for the first one) and `fn give_suggestion_to_make_binding_mutable()` (for the second one). Also some comments, ideally with code examples, is also helpful in these kinds of situations (or at least links to test cases) -- basically explaining "if the input program is this, we want this error"